doc/module-system/module-system.chapter.md

```diff
@@ -1,14 +1,207 @@
+- **documentation**, so that each option has a description, conveniently located in the file that declares it
+- **modularity** or **aspect oriented programming**: separate files can all contribute to the same option.
+
+The following sections are dense introductory reference material.
+
+<!-- TODO: link the tutorial
+   If you're learning to work with the module system, consider following the introductory tutorial to get a feel for the module system. -->
+
+## Concepts {#module-system-concepts}
+
+The following sections briefly define the concepts that make up the module system.
+
+<!-- keep them short -->
+
+### Module Evaluation {#module-system-concept-evaluation}
+
+<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->
+
+Module evaluation is the process of turning a set of modules into option values.
+
+The module system can be seen as an embedded domain specific language (eDSL) for configuration. The defining feature of an eDSL is that they are evaluated by a host language, which in this case is the Nix expression language. So _module evaluation_ is implemented using Nix evaluation, by means of the [`evalModules`] function.
+
+### Configuration {#module-system-concept-configuration}
+
+<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->
+
+A configuration is the result of [module evaluation](#module-system-concept-evaluation).
+
+It consists of a set of options. Each option is identified by an [attribute path], and may produce a single value (which may in turn be a composite value such as an attribute set).
+
+The module system provides its caller with _all_ the options and their values. It is up to the [application] to assign meaning to the options and to pick options (typically one option) that represents the value of the entire configuration.
+
+### Module {#module-system-concept-module}
+
+<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->
+
+A module is a function that produces part of the configuration.
+
+A module can be written using multiple, similar [syntaxes](#module-syntax), which are converted internally to the most general syntax, which is a function that returns an attribute set with specific attribute names.
+
+These attributes, the [module attributes] are then processed in order to produce the [module arguments], which are passed to the modules.
+
+As you can see, module system evaluation is a cyclic process. This works thanks to Nix's lazy evaluation.
+
+### Data flow {#module-system-data-flow}
+
+<!-- NOTE: this is an introductory section. It is not meant to be a complete reference -->
+
+The following diagram shows the data flows that occur within a module system evaluation.
+
+Multiplicity is not modeled. For example `.options -> options` is a many-to-many transformation where any option declaration can affect any option.
+
+```mermaid
+flowchart TD
```

Comment:
The nodes could have links to anchors in the text to provide more details on what each one is for and what it means.